src/main/webapp/WEB-INF/view/activityfeed.jsp

```diff
@@ -28,17 +38,40 @@
     <p>Here's everything that's happened on the site so far!</p>
 
     <div id="activity">
-      <%List<User> users = 
-        (List<User>) request.getAttribute("users");
-      %>
       <ul>
-        <% 
-          for(User user : users) {
-            Date date = Date.from(user.getCreationTime());
+        <%
+        for(Map.Entry<Instant, Object> entry : sortedEventsMap.entrySet()) {
+          Date date = Date.from(entry.getKey());
         %>
-        <li><b><%= date %>: </b> 
-          <a href="/profile/<%= user.getName() %>">
-          <%= user.getName() %></a> joined!</li>
+          <li>
+            <b><%= date %>:</b>
+          <%
+          if(entry.getValue() instanceof User) {
```

Comment:
An alternative implementation to consider is to build a new class, let's call it `Event`, that contains details such as the type of event, date, the username if it's a join event, the conversation id and author if it's a new conversation.. etc, then passing it to the jsp, which uses it directly to render the HTML.

Reply:
I'll try this implementation, it will be a lot cleaner. Thanks!